synapse/handlers/room_member.py

```diff
@@ -203,15 +202,19 @@ def _local_membership_update(
             # info.
             newly_joined = True
             if prev_member_event_id:
-                prev_member_event = yield self.store.get_event(prev_member_event_id)
+                prev_member_event = await self.store.get_event(prev_member_event_id)
                 newly_joined = prev_member_event.membership != Membership.JOIN
             if newly_joined:
-                yield self._user_joined_room(target, room_id)
+                d = self._user_joined_room(target, room_id)
+                if d:
+                    await d
```

Comment:
I think we talked a bit about this in #synapse-dev:matrix.org, but don't fully remember the result. When does this return false-y?

The implementation here is a bit confusing to follow so not sure if I found all the proper instances of `_user_joined_room`.

Reply:
I thought these were the two implementations:

https://github.com/matrix-org/synapse/blob/3eab76ad43e1de4074550c468d11318d497ff632/synapse/handlers/room_member.py#L1008-L1011

https://github.com/matrix-org/synapse/blob/3eab76ad43e1de4074550c468d11318d497ff632/synapse/handlers/room_member_worker.py#L71-L76

How does the notifier code come into play? (Granted I'm not sure I understand how that distributor thing works.)

Reply:
We've decided to split this out into a separate PR as at first glance it doesn't look like a simple change.

Reply:
f7264e63199c6653610b848f5f5ddd8e78e0ab3b is a decent half-way measure for this.